homeassistant/components/roborock/__init__.py

```diff
@@ -29,66 +44,140 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-            mqtt_clients[device.duid],
-        )
-    await asyncio.gather(
-        *(coordinator.verify_api() for coordinator in coordinator_map.values())
-    )
-    # If one device update fails - we still want to set up other devices
-    await asyncio.gather(
+    # Get a Coordinator if the device is available or if we have connected to the device before
+    coordinators = await asyncio.gather(
         *(
-            coordinator.async_config_entry_first_refresh()
-            for coordinator in coordinator_map.values()
-        ),
-        return_exceptions=True,
+            setup_device(
+                hass,
+                user_data,
+                device,
+                product_info[device.product_id],
+                CachedCoordinatorInformation(
+                    network_info=NetworkInfo(
+                        **entry.data[CONF_CACHED_INFORMATION][device.duid][
+                            "network_info"
+                        ]
+                    ),
+                    supported_entities=set(
+                        entry.data[CONF_CACHED_INFORMATION][device.duid][
+                            "supported_entities"
+                        ]
+                    ),
+                )
+                if device.duid in entry.data[CONF_CACHED_INFORMATION]
+                else None,
+            )
+            for device in device_map.values()
+        )
     )
+    # Valid coordinators are those where we had networking cached or we could get networking
+    valid_coordinators: list[RoborockDataUpdateCoordinator] = [
+        coord for coord in coordinators if coord is not None
+    ]
+    if len(valid_coordinators) == 0:
+        raise ConfigEntryNotReady("No coordinators were able to successfully setup.")
     hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
-        device_id: coordinator
-        for device_id, coordinator in coordinator_map.items()
-        if coordinator.last_update_success
-    }  # Only add coordinators that succeeded
+        coordinator.roborock_device_info.device.duid: coordinator
+        for coordinator in valid_coordinators
+    }
 
     if not hass.data[DOMAIN][entry.entry_id]:
         # Don't start if no coordinators succeeded.
         raise ConfigEntryNotReady("There are no devices that can currently be reached.")
-
+    updated_cached: dict[str, dict] = {}
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
-
+    for coord in valid_coordinators:
+        updated_cached.update(
+            {
+                coord.roborock_device_info.device.duid: asdict(
+                    CachedCoordinatorInformation(
+                        network_info=coord.roborock_device_info.network_info,
+                        supported_entities=coord.supported_entities,
+                    )
+                )
+            }
+        )
+    hass.config_entries.async_update_entry(
+        entry, data={**entry.data, CONF_CACHED_INFORMATION: updated_cached}
```

Comment:
After this PR - I plan to add dhcp discovery, and knowledge about what devices current ip addresses are would be helped - so I added it here while I was already changing the config flow, to avoid having to do two migrations. I could remove it though if you would like

Reply:
oh actually jk - network information is needed to setup a local client - I forgot I added that capability here.

Reply:
Sorry for the delay. I asked internally about storing cached information, and the conclusion is:

Config entries are not a generic caching/data storage.
It's fine to have a known/limited list with connection details from a flow in the config entry.
But for the rest, you should create your storage.

For now, please remove it, as you don't need any cached information in any flow.

As you are gathering this information on each setup, maybe you could store it in `hass.data`? Otherwise please create your own storage

Reply:
Do you happen to have any good examples that implement caching in a separate storage - is just basic file io what is preferred here? I also though hass.data didn't persist on restart, but I guess that was just a misunderstanding on my part. 

Once I add dhcp discovery - I will need the networking information that is cached to ensure it is a new device - but I could do that with the extra storage as well I assume

Reply:
Changed to use Store() Let me know if that is okay

Reply:
My question is, why do you need this information across restarts? After a restart, you can trust the mac address but not the IP address as that could be changed in the meantime

Reply:
The reasoning is that users will sometimes not be able to access the cloud server because they are rate limited, or there is a problem with their account or they have some other issue getting the networking. basically - we want to attempt to get the information from the cloud, but if we cannot, we should try previous information.

Reply:
And having this information saved means we can still create the device and just disable it rather than not creating the device at all, which was the start of our discussion